components/feature/downloads/src/main/java/mozilla/components/feature/downloads/DownloadMiddleware.kt

```diff
@@ -101,7 +101,7 @@ class DownloadMiddleware(
         downloadStorage.getDownloads().collect { downloads ->
             downloads.forEach { download ->
                 if (!context.state.downloads.containsKey(download.id)) {
-                    context.dispatch(DownloadAction.RestoreDownloadStateAction(download))
+                    context.store.dispatch(DownloadAction.RestoreDownloadStateAction(download))
```

Comment:
This was the quick fix. But I'd recommend not passing the `MiddlewareContext` to this Coroutine at all since accessing it is not correct in any case.

Reply:
Thanks for the feedback, could you further explain to understand better the issue?, what approach do you recommend in this case?

Reply:
pass the result of `context.store`, a `BrowserStore` around to coroutines like here and not the context object. Just to avoid using the wrong methods completely. :)

Reply:
Something like this?
```kotlin
    private fun restoreDownloads(context: MiddlewareContext<BrowserState, BrowserAction>) {
        val store = context.store
        scope.launch {
            downloadStorage.getDownloads().collect { downloads ->
                downloads.forEach { download ->
                    if (!store.state.downloads.containsKey(download.id)) {
                        store.dispatch(DownloadAction.RestoreDownloadStateAction(download))
                        logger.debug("Download restarted from the storage ${download.fileName}")
                    }
                }
            }
        }
    }
```

Reply:
I'd move this earlier to the call `restoreDownloads(context.store)`, then you have no way of accessing `context` inside the `launch` block :)

Reply:
That's even nicer!!!

Reply:
@pocmo @Amejia481 but with this pattern we have the inverse problem. If we pass the store around in the middleware then devs tend to call `store.dispatch` when they really want the synchronous `context.dispatch` in the middleware. (That happened last week too)

Maybe we should rename the latter to `syncDispatch` to make this more obvious.

Reply:
Oh yeah, I only meant in this specific case where everything the function does is inside a coroutine.

```Kotlin
private fun restoreDownloads(...) = scope.launch {
  ...
}
```